src/extensions/arith.rs

```diff
@@ -0,0 +1,1080 @@
+//! Miniscript Arithmetic expressions:
+//! Note that this fragment is only supported for Tapscript context
+use std::str::FromStr;
+use std::{cmp, error, fmt};
+
+use elements::opcodes::all::*;
+use elements::sighash::Prevouts;
+use elements::{opcodes, script, Transaction};
+
+use super::{ExtParam, ParseableExt};
+use crate::expression::{FromTree, Tree};
+use crate::miniscript::context::ScriptContextError;
+use crate::miniscript::lex::{Token as Tk, TokenIter};
+use crate::miniscript::limits::MAX_STANDARD_P2WSH_STACK_ITEM_SIZE;
+use crate::miniscript::satisfy::{Satisfaction, Witness};
+use crate::miniscript::types::extra_props::{OpLimits, TimelockInfo};
+use crate::miniscript::types::{Base, Correctness, Dissat, ExtData, Input, Malleability};
+use crate::{
+    expression, interpreter, miniscript, script_num_size, Error, ExtTranslator, Extension,
+    Satisfier, ToPublicKey, TranslateExt,
+};
+
+/// Enum representing arithmetic operations with transaction amounts.
+/// Every variant of this enum pushes a single singed 64 bit BE number on stack top.
+/// All of introspection opcodes explicitly assert the amount is explicit.
+///
+/// This will abort when
+///     - Any of operations are on confidential amounts. The Null case is automatically
+///       converted to explicit zero.
+///     - Supplied index is out of bounds.
+///     - Any of the operations overflow. Refer to tapscript opcodes spec for overflow specification
+///     - In extreme cases, when recursive operations exceed 400 depth
```

Comment:
In 97f16a908cb974e4001308a2418cef6e32130385:

Because we use recursive algorithms everywhere I'm unsure that we'll actually hit 400 depth without crashing the library ... something to revisit in a later PR.